esp-hal/src/i2c/master/mod.rs

```diff
@@ -149,6 +149,86 @@ impl BusTimeout {
     }
 }
 
+#[cfg(not(esp32))]
+cfg_if::cfg_if! {
+    if #[cfg(esp32s2)] {
+        type FsmTimeoutType = u32;
+    } else {
+        type FsmTimeoutType = u8;
+    }
+}
+
+/// When the FSM remains unchanged for more than
+#[cfg_attr(esp32s2, doc = "the given amount of bus clock cycles")]
+#[cfg_attr(not(esp32), doc = "the 2^ the given amount of bus clock cycles")]
+/// a timeout will be triggered.
+///
+/// The default value is
+#[cfg_attr(esp32s2, doc = "0x100")]
+#[cfg_attr(not(esp32), doc = "0x10")]
+#[instability::unstable]
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+#[cfg(not(esp32))]
+pub struct FsmTimeout {
+    value: FsmTimeoutType,
+}
+
+#[cfg(not(esp32))]
+impl FsmTimeout {
+    cfg_if::cfg_if! {
+        if #[cfg(esp32s2)] {
+            const FSM_TIMEOUT_MAX: FsmTimeoutType = 0xff_ffff;
+        } else {
+            const FSM_TIMEOUT_MAX: FsmTimeoutType = 23;
+        }
+    }
+
+    cfg_if::cfg_if! {
+        if #[cfg(esp32s2)] {
+            const FSM_TIMEOUT_DEFAULT: FsmTimeoutType = 0x100;
+        } else {
+            const FSM_TIMEOUT_DEFAULT: FsmTimeoutType = 0x10;
+        }
+    }
+
+    /// Creates a new timeout.
+    ///
+    /// The meaning of the value and the allowed range of values is different
+    /// for different chips.
+    pub const fn new_const<const VALUE: FsmTimeoutType>() -> Self {
+        const {
+            core::assert!(VALUE <= Self::FSM_TIMEOUT_MAX, "Invalid timeout value");
+        }
+        Self { value: VALUE }
+    }
+
+    /// Creates a new timeout.
+    ///
+    /// The meaning of the value and the allowed range of values is different
+    /// for different chips.
+    pub fn new(value: FsmTimeoutType) -> Result<Self, ConfigError> {
+        if value > Self::FSM_TIMEOUT_MAX {
+            return Err(ConfigError::TimeoutInvalid);
+        }
+
+        Ok(Self { value })
+    }
+
+    fn value(&self) -> FsmTimeoutType {
+        self.value
+    }
+}
+
+#[cfg(not(esp32))]
+impl Default for FsmTimeout {
+    fn default() -> Self {
+        Self {
```

Comment:
We should probably use Self::new_const here, just to make sure we don't accidentally break the default value.